### PR TITLE
Replace dfu-util with particle usb dfu

### PR DESCRIPTION
**Pull request by keeramis**
## Description

This PR removes the last uses of `dfu-util` to use `particle usb dfu` for improved compatibility.

### Other improvements:
1. Uses usbUtils to handle multiple devices connected simultaneously
2. Refactors promise-based methods in the keys module to utilize the `async/await` syntax, improving code readability and maintainability.
3. Removes the `particle keys protocol` commands

## How to Test

This PR requires that the latest `particle usb` upload functionality exists which currently is in this PR.
Checkout the `particle-usb` repo locally and link your locally checked out `particle-cli` to use locally checked out `particle-usb`

1. Checkout this branch locally and follow the instructions in the README (run `nvm use` followed by `npm install`)
2. Checkout [this branch](https://github.com/particle-iot/particle-usb/tree/feature/impl_dfu_upload) of `particle-usb` locally. 
    1. Go to `particle-usb` root directory. Run `npm link`. Ensure it is linked as expected
    2. Go to `particle-cli` root directory. Run `npm link particle-usb`
    3. Verify from `particle-cli/node_modules` that the `particle-usb` points to your local repo

5. Run any `particle keys xxx` commands and `particle doctor` command on your device(s) and verify that they are working as expected. 
6. For example, To run `particle keys xxx`, run `npm start -- keys xxx` from project's root directory

Bonus points if you are able to verify that the keys are generated as expected!

## Related Issues / Discussions

Particle-USB PR that is needed for this PR to work: https://github.com/particle-iot/particle-usb/pull/91 

- https://app.shortcut.com/particle/story/121835/remove-last-uses-of-dfu
- https://app.shortcut.com/particle/story/122385/remove-usage-of-dfu-util-in-particle-cli
- https://app.shortcut.com/particle/story/122384/use-async-await-instead-of-promises-for-keys-module

## Completeness

- [x] User is totes amazing for contributing!
- [ ] Contributor has signed [CLA](https://docs.google.com/a/particle.io/forms/d/1_2P-vRKGUFg5bmpcKLHO_qNZWGi5HKYnfrrkd-sbZoA/viewform)
- [ ] Problem and solution clearly stated
- [ ] Tests have been provided
- [ ] Docs have been updated
- [ ] CI is passing

### CODE REVIEW

File: src/cmd/doctor.js [hunk 1]
```diff
@@ -67,14 +68,38 @@ module.exports = class DoctorCommand {
...
+		// TODO: Replace the serial commands to use usb-utils
+
 		// Try to find a "normal" mode device through the serial port
 		return this.command('serial')
 			.findDevices()
 			.then(devices => {
 				if (devices.length === 0){
-					// Try to find a "DFU" mode device through dfu-util
-					return dfu.listDFUDevices();
+					return this.getDfuDevices();
```
Comment (keeramis) [inline src/cmd/doctor.js:102]: I replaced the `dfu.listDFUDevices` to use usbUtils in such a way that this line exactly returns the object that was previously returned by dfu.listDFUDevices.

Pasting the return values from 
--> getDfuDevices of this PR
[getDfuDevicesRes Full output](https://github.com/particle-iot/particle-cli/files/12738191/getDfuDevicesRes.txt)

```
[
  {
    "type": "Photon 2 / P2",
    "dfuId": "2b04:d020",
    "specs": {
      .....
      }
    }
  },
  {
    "type": "Argon",
    "dfuId": "2b04:d00c",
    "specs": {
      ......
      }
    }
  }
]
```

---> From dfu.listDFUDevices [listDFUDevices full output](https://github.com/particle-iot/particle-cli/files/12738310/listDFUDevicesResp.txt)

```
[
  {
    "type": "Argon",
    "dfuId": "2b04:d00c",
    "specs": {
      .....
    }
  },
  {
    "type": "Photon 2 / P2",
    "dfuId": "2b04:d020",
    "specs": {
     ....
    }
  }
]
```

**Comment by keeramis** [comment]
It's worth noting an interesting observation related to 'particle keys.' We create a key of 121 bytes. However, when we save this key, we save 192 bytes. Will it be beneficial to consider optimizing this process to save only the original 121 bytes, or alternatively, to provide documentation explaining this behavior.

File: src/cmd/keys.js [hunk 2]
```diff
@@ -21,208 +24,154 @@ const dfu = require('../lib/dfu');
...
+			protocol = await this.validateDeviceProtocol({ device });
+			let segmentName = this._getPrivateKeySegmentName({ protocol });
+			let segment = this._validateSegmentSpecs(segmentName);
+			let buf;
+			try {
+				buf = await device.readOverDfu({ altSetting: segment.specs.alt, startAddr: segment.specs.address, size: segment.specs.size });
+			} catch (err) {
+				// FIXME: First time read may fail so we retry
+				buf = await device.readOverDfu({ altSetting: segment.specs.alt, startAddr: segment.specs.address, size: segment.specs.size });
+			}
```
Comment (keeramis) [inline src/cmd/keys.js:146]: As mentioned, there is a challenge when calling `writeOverDfu` and it is followed by a `readOverDfu`. I noticed that there is a USB Stall Error during the first read operation which gets cleared during the second read operation. I am currently debugging this but added a workaround here to temporarily unblock this issue.

File: src/cmd/keys.js [hunk 3]
```diff
@@ -677,5 +610,56 @@ module.exports = class KeysCommand {
...
+
+	async _dfuRead(device, { altSetting, startAddr, size }) {
+		let buf;
+		try {
+			buf = await device.readOverDfu({ altSetting, startAddr, size });
+		} catch (err) {
+			// FIXME: First time read may fail so we retry
+			buf = await device.readOverDfu({ altSetting, startAddr, size });
+		}
+		return buf;
```
Comment (keeramis) [inline src/cmd/keys.js:655]: Open to improving this logic for more accurate error detection just in case the 'readOverDfu' operation fails again.

**Review by monkbroc — APPROVED** [review]

**Pull request merged**